python_modules/dagster/dagster/_config/pythonic_config/__init__.py

```diff
@@ -812,16 +825,56 @@ def with_resource_context(
         return copy
 
     def _initialize_and_run(self, context: InitResourceContext) -> TResValue:
-        updated_resource = (
-            self._resolve_and_update_nested_resources(context)  # noqa: SLF001
-            .with_resource_context(context)
-            ._resolve_and_update_env_vars()
-        )
+        with self._resolve_and_update_nested_resources(context) as has_nested_resource:
+            updated_resource = has_nested_resource.with_resource_context(  # noqa: SLF001
+                context
+            )._resolve_and_update_env_vars()
 
-        return updated_resource._create_object_fn(context)  # noqa: SLF001
+            updated_resource.pre_execute(context)
+            return updated_resource.create_resource(context)
 
-    def _create_object_fn(self, context: InitResourceContext) -> TResValue:
-        return self.create_resource(context)
+    @contextlib.contextmanager
+    def _initialize_and_run_cm(
+        self, context: InitResourceContext
+    ) -> Generator[TResValue, None, None]:
+        with self._resolve_and_update_nested_resources(context) as has_nested_resource:
+            updated_resource = has_nested_resource.with_resource_context(  # noqa: SLF001
+                context
+            )._resolve_and_update_env_vars()
+
+            with updated_resource.yield_for_execution(context) as value:
+                yield value
+
+    def pre_execute(self, context: InitResourceContext) -> None:
+        """Optionally override this method to perform any pre-execution steps
+        needed before the resource is used in execution.
+        """
+        pass
+
+    def post_execute(self, context: InitResourceContext) -> None:
+        """Optionally override this method to perform any post-execution steps
+        needed after the resource is used in execution.
+
+        post_execute will be called even if the resource fails to initialize in the
+        pre_execute step, and if any part of the run fails.
+        """
+        pass
+
+    @contextlib.contextmanager
+    def yield_for_execution(self, context: InitResourceContext) -> Generator[TResValue, None, None]:
+        """Optionally override this method to perform any lifecycle steps
+        before or after the resource is used in execution. By default, calls
+        pre_execute before yielding, and post_execute after yielding.
+
+        Note that if you override this method and want pre_execute or
+        post_execute to be called, you must invoke them yourself.
+        """
+        try:
+            self.pre_execute(context)
```

Comment:
Need to talk through the decision about whether `pre_execute` is in the `try` or not.

If this is the right decision, then we should change the name of the methods as they would now be misleading. Namely we would have to write a sentence like "if pre_execute throws, then post_execute will be called, even though no execution has occured".

We also need to think ahead in case we develop a feature to execute code for a resource on run start, rather than for step start.

Most explicit and accuarate:

`pre_execute` --> `setup_for_step_execution`
`post_execute` --> `teardown_after_step_execution`

In this naming, we should say: "If setup_for_step_execution fails, then step execution is failed. As a result `teardown_after_step_execution` will be called."

Reply:
I don't think `setup_for_execution_process` as the word "process" itself is ambiguous between the "operating system process" and "general process".

`setup_for_execution` and `teardown_after_execution` seem good to me

Reply:
On the face of it, those names sound good to me. I think it's worth considering how they'd sound if we eventually decide to also add run-level setup / teardown methods (e.g. for setting up / tearing down a cluster per run).

Reply:
Yeah agree sandy. I commented similar above:

> We also need to think ahead in case we develop a feature to execute code for a resource on run start, rather than for step start.

I'm also pretty confident we won't want to make to them methods, lest the model for state in these classes gets even _more_ complicated, but instead some other lifecycle mechanism that accepts the resource as an argument much more akin to an `op`. Otherwise the user will have to thing of the object in a bunch in a huge number of states. (e.g. I am in process where the  `setup_for_execution` method has been called, but not the `on_run_start` call, since this is a process that is in the middle of run). 

We can dig into that in another PR/discussion.

Reply:
> I'm also pretty confident we won't want to make to them methods, lest the model for state in these classes gets even more complicated

I agree that it shouldn't be able to modify the state of the object. Maybe a static method?